include/dice/template-library/type_list.hpp

```diff
@@ -0,0 +1,413 @@
+    struct all_of;
+
+    template<typename ...Ts, auto pred>
+    struct all_of<type_list<Ts...>, pred> : std::bool_constant<(std::invoke(pred, std::type_identity<Ts>{}) && ...)> {
+    };
+
+    template<typename TL, auto pred>
+    inline constexpr bool all_of_v = all_of<TL, pred>::value;
+
+
+    /**
+     * Check if any of the types in the type list fulfill a predicate.
+     *
+     * @tparam TL type list
+     * @tparam pred predicate to apply to types. Must be invocable as INVOKE<bool>(pred, std::type_identity<T>{}) for all types in the type list.
+     */
+    template<typename TL, auto pred>
+    struct any_of;
+
+    template<typename ...Ts, auto pred>
+    struct any_of<type_list<Ts...>, pred> : std::bool_constant<(std::invoke(pred, std::type_identity<Ts>{}) || ...)> {
+    };
+
+    template<typename TL, auto pred>
+    inline constexpr bool any_of_v = any_of<TL, pred>::value;
+
+
+    /**
+     * Check if no types in the type list fulfill a predicate.
+     *
+     * @tparam TL type list
+     * @tparam pred predicate to apply to types. Must be invocable as INVOKE<bool>(pred, std::type_identity<T>{}) for all types in the type list.
+     */
+    template<typename TL, auto pred>
+    struct none_of;
+
+    template<typename ...Ts, auto pred>
+    struct none_of<type_list<Ts...>, pred> : std::bool_constant<(!std::invoke(pred, std::type_identity<Ts>{}) && ...)> {
+    };
+
+    template<typename TL, auto pred>
+    inline constexpr bool none_of_v = none_of<TL, pred>::value;
+
+
+	/**
+     * Check if all types in the type list are the same type.
+     *
+     * @tparam TL type list
+     */
+    template<typename TL>
+    struct all_same;
+
+    template<>
+    struct all_same<type_list<>> : std::true_type {
+    };
+
+    template<typename T, typename ...Ts>
+    struct all_same<type_list<T, Ts...>> : std::bool_constant<(std::is_same_v<T, Ts> && ...)> {
+    };
+
+    template<typename TL>
+    inline constexpr bool all_same_v = all_same<TL>::value;
+
+
+	/**
+     * A marker type to indicate a failed operation
+     * when using opt
+     */
+    struct nullopt {};
+
+	/**
+     * Can be used in combination with any operation that may fail.
+     * By default, failing operations will not provide a `type` member typedef.
+     * When using opt, the member typedef `type` will always be present. In case the operation failed the type will be `nullopt`.
+     *
+     * @example
+     * @code
+     * using X = first_t<type_list<>>; // compilation failure, first has no `type` typedef
+     * using X = opt_t<first<type_list<>>; // compilation success, X is nullopt
+     * @endcode
+     */
+    template<typename T, typename = void>
+    struct opt {
+        using type = nullopt;
+    };
+
+    template<typename T>
+    struct opt<T, std::void_t<typename T::type>> {
+        using type = typename T::type;
+    };
+
+    template<typename T>
+    using opt_t = typename opt<T>::type;
+
+
+    namespace detail_for_each_fold {
```

Comment:
do they need to be in the same namespace? if not, i suggest to split this into two.